File: matrix.go
/*
参考URL: https://qiita.com/hamadu/items/fce4ee1e4b5c2c2d24df
逐次的に行列演算を行う
第一引数として，行列のサイズを指定する
*/

package main

import (
	"fmt"
	"os"
	"strconv"
	"time"
)

type Matrix [][]int

func mul(a, b *Matrix) Matrix {
	ar := len(*a)
	ac := len((*a)[0])
	br := len(*b)
	bc := len((*b)[0])

	// 縦横のサイズが合わない場合
	if ac != br {
		panic("wrong matrix type")
	}

	// ここが O(n^3) になってる
	c := make(Matrix, ar)
	for i := 0; i < ar; i++ {
		c[i] = make([]int, bc)
		for j := 0; j < bc; j++ {
			for k := 0; k < ac; k++ {
				c[i][j] += (*a)[i][k] * (*b)[k][j]
			}
		}
	}
	return c
}

func main() {
	n, _ := strconv.Atoi(os.Args[1])

	a := make(Matrix, n)
	b := make(Matrix, n)

	//値50を行列に格納
	for i := 0; i < n; i++ {
		a[i] = make([]int, n)
		b[i] = make([]int, n)
		for j := 0; j < n; j++ {
			a[i][j] = 50
			b[i][j] = 50
		}
	}

	var ans Matrix
	start := time.Now() //測定開始

	ans = mul(&a, &b)

	end := time.Now() //測定終了
	exeSpeed := end.Sub(start)
	fmt.Printf("time = %.10f sec \n", exeSpeed.Seconds())
	fmt.Println(ans)
}

File: Matrix/Ruby/matrix.rb
class Array
    def mul(x)
      m = self.size - 1
      n = x.size - 1
      c = Array.new(m + 1)
      c.each_index {|i| c[i] = Array.new(m + 1, 0)}
      for i in 0..m
        for j in 0..m
      for k in 0..n
        c[i][j] += self[i][k] * x[k][j]
      end
        end
      end
      c
    end
  end
  
  n =ARGV[0].to_i
  m =ARGV[0].to_i

  start_time = Time.now
  m-=1
  n-=1
  a = Array.new(m + 1)
  a.each_index {|i| a[i] = Array.new(m + 1, 0)}
  b = Array.new(m + 1)
      b.each_index {|i| b[i] = Array.new(m + 1, 0)}
  for i in 0..m
    for j in 0..n
        a[m][n]=50
        b[m][n]=50
    end
  end
   a.mul(b)  #=>[[-1, -3, 5], [-4, 6, -4], [5, -3, -1]]
  p "TIME #{Time.now - start_time}s"